internal/remote/client_factory.go

```diff
@@ -0,0 +1,107 @@
+package remote
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	apicorev1 "k8s.io/api/core/v1"
+	apimetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8slabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kyma-project/lifecycle-manager/api/shared"
+)
+
+type SkrContextFactory interface {
+	Get(kyma types.NamespacedName) (*SkrContext, error)
+	Init(ctx context.Context, kyma types.NamespacedName) error
+}
+
+type KymaSkrContextFactory struct {
+	clientCache *ClientCache
+	kcpClient   Client
+}
+
+func NewKymaSkrContextFactory(kcpClient Client, clientCache *ClientCache) *KymaSkrContextFactory {
+	return &KymaSkrContextFactory{
+		clientCache: clientCache,
+		kcpClient:   kcpClient,
+	}
+}
+
+const KubeConfigKey = "config"
+
+var (
+	ErrAccessSecretNotFound     = errors.New("access secret not found")
+	ErrSkrClientContextNotFound = errors.New("skr client context not found")
+)
+
+func (k *KymaSkrContextFactory) Init(ctx context.Context, kyma types.NamespacedName) error {
+	if k.clientCache.Contains(kyma) {
+		return nil
+	}
+
+	kubeConfigSecretList := &apicorev1.SecretList{}
+	if err := k.kcpClient.List(ctx, kubeConfigSecretList, &client.ListOptions{
+		LabelSelector: k8slabels.SelectorFromSet(k8slabels.Set{shared.KymaName: kyma.Name}), Namespace: kyma.Namespace,
+	}); err != nil {
+		return fmt.Errorf("failed to list kubeconfig secrets: %w", err)
+	} else if len(kubeConfigSecretList.Items) < 1 {
+		return fmt.Errorf("secret with label %s=%s %w", shared.KymaName, kyma.Name, ErrAccessSecretNotFound)
+	}
+
+	kubeConfigSecret := kubeConfigSecretList.Items[0]
+
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(kubeConfigSecret.Data[KubeConfigKey])
+	if err != nil {
+		return fmt.Errorf("failed to create rest config from kubeconfig: %w", err)
+	}
+
+	restConfig.QPS = k.kcpClient.Config().QPS
+	restConfig.Burst = k.kcpClient.Config().Burst
+
+	clnt, err := client.New(restConfig, client.Options{Scheme: k.kcpClient.Scheme()})
+	if err != nil {
+		return fmt.Errorf("failed to create lookup client: %w", err)
+	}
+
+	skrClient := NewClientWithConfig(clnt, restConfig)
+
+	namespace := &apicorev1.Namespace{
```

Comment:
this whole "ensures remote namespace exists" logic can be extracted to the main reconciler logic, then you don't need to implement such logic in `DualClusterFactory`, `SingleClusterFactory` for the integration test. It's not a logic of this init function should care about.

Reply:
something like this in kyma reconciler
```
if r.SyncKymaEnabled(kyma) {
    err := r.SkrContextFactory.Init(ctx, kyma.GetNamespacedName())
    skrContext, err := r.SkrContextFactory.Get(kyma.GetNamespacedName())
    if err != nil {
	    return ctrl.Result{}, err
    }
    if err := r.ensureRemoteNamespaceExits(ctx, skrContext); err != nil {
	    return ctrl.Result{}, err
    }
...
```